csharp/src/Ice/Internal/ObserverMiddleware.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) ZeroC, Inc.
+
+#nullable enable
+
+using Ice.Instrumentation;
+
+namespace Ice.Internal;
+
+/// <summary>A middleware that observes dispatches.</summary>
+internal sealed class ObserverMiddleware : Object
+{
+    public ValueTask<OutgoingResponse> dispatchAsync(IncomingRequest request)
+    {
+        return _communicatorObserver.getDispatchObserver(request.current, request.size) is DispatchObserver observer ?
+            performDispatchAsync(observer) :
+            _next.dispatchAsync(request);
+
+        async ValueTask<OutgoingResponse> performDispatchAsync(DispatchObserver observer)
+        {
+            observer.attach();
+            try
+            {
+                OutgoingResponse response;
+                try
+                {
+                    response = await _next.dispatchAsync(request).ConfigureAwait(false);
+                }
+                catch (System.Exception ex)
+                {
+                    // We need to marshal the exception into the response immediately to observe the response size.
+                    response = request.current.createOutgoingResponse(ex);
```

Comment:
Shouldn't we rethrow the exception?

Reply:
No.

The whole logic of this code is to marshal the exception into the response. And as explained by the comment on line 30, it's because we need to "observe" the size of the response.